backend/service/chaos/experimentation/terminator/terminator.go

```diff
@@ -0,0 +1,176 @@
+	perExperimentCheckInterval time.Duration
+
+	log *zap.SugaredLogger
+
+	activeMonitoringRoutines trackingGauge
+	terminationCount         tally.Counter
+	marshallingErrors        tally.Counter
+}
+
+func (m *monitor) Run(ctx context.Context) {
+	for _, configType := range m.enabledConfigTypes {
+		// For each monitored config type, start a single goroutine that polls the active experiments at a fixed interval.
+		// Whenever a new (new to this goroutine) experiment is found, open up a goroutine that periodically evaluates all
+		// the termination criteria for the experiment.
+		//
+		// This approach ensures provides a steady DB pressure (mostly outer loop, some from triggering termination) and relatively high
+		// fairness, as checking the termination conditions for one experiment should not be delaying the checks for another experiment
+		// unless we're under very heavy load.
+		go func(configType string) {
+			trackedExperiments := map[uint64]context.CancelFunc{}
+			ticker := time.NewTicker(m.outerLoopInterval)
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					es, err := m.store.GetExperiments(context.Background(), configType, experimentationv1.GetExperimentsRequest_STATUS_RUNNING)
+					if err != nil {
+						m.log.Errorw("failed to retrieve experiments from experiment store", "err", err, "enableConfigTypes", m.enabledConfigTypes)
+						continue
+					}
+
+					activeExperiments := m.monitorNewExperiments(es, trackedExperiments)
+
+					// For all experiments that we're tracking that no longer appear to be active, cancel the goroutine
+					// and clean it up.
+					for e, cancel := range trackedExperiments {
+						if _, ok := activeExperiments[e]; !ok {
+							cancel()
+							delete(trackedExperiments, e)
+						}
+					}
+				}
+			}
+		}(configType)
+	}
+}
+
+// monitorNewExperiments iterates over all the provided experiments, spawning a goroutine to montior each experiment that
+// doesn't already have a monitoring routine. Returns a set containing all the active experiment ids for further processing.
+func (m *monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trackedExperiments map[uint64]context.CancelFunc) map[uint64]struct{} {
+	// For each active experiment, create a monitoring goroutine if necessary.
+	activeExperiments := map[uint64]struct{}{}
+	for _, e := range es {
+		activeExperiments[e.Id] = struct{}{}
+		if _, ok := trackedExperiments[e.Id]; !ok {
+			ctx, cancel := context.WithCancel(context.Background())
+			trackedExperiments[e.Id] = cancel
+
+			m.activeMonitoringRoutines.inc()
+			go func() {
+				defer m.activeMonitoringRoutines.dec()
+				m.monitorSingleExperiment(ctx, e)
+			}()
+		}
+	}
+	return activeExperiments
+}
+
+func (m *monitor) monitorSingleExperiment(ctx context.Context, e *experimentationv1.Experiment) {
+	ticker := time.NewTicker(m.perExperimentCheckInterval)
+	terminated := false
+
+	dany := &ptypes.DynamicAny{}
```

Comment:
so this can be just casted to a specific type - i.e., serverexperimentation model - without any conversions?

Reply:
Updated to use a proto.Message as per the deprecation warnings on DynamicAny, so yeah this can just be casted to the specific proto

Reply:
nice, created an internal ticket for changing the implementation of `ExperimentConfig` so that it uses `DynamicAny`